rhizome/lib/vm_setup.rb

```diff
@@ -73,17 +67,17 @@ def purge
   end
 
   def interfaces
-    r "ip netns add #{q_vm}"
+    r "ip netns list | grep -q '^#{q_vm} ' || ip netns add #{q_vm}"
 
     # Generate MAC addresses rather than letting Linux do it to avoid
     # a vexing bug whereby a freshly created link will, at least once,
     # spontaneously change its MAC address sometime soon after
     # creation, as caught by instrumenting reads of
     # /sys/class/net/vethi#{q_vm}/address at two points in time.  The
     # result is a race condition that *sometimes* worked.
-    r "ip link add vetho#{q_vm} addr #{gen_mac.shellescape} type veth peer name vethi#{q_vm} addr #{gen_mac.shellescape} netns #{q_vm}"
+    r "ip link show vetho#{q_vm} > /dev/null 2>&1 || ip link add vetho#{q_vm} addr #{gen_mac.shellescape} type veth peer name vethi#{q_vm} addr #{gen_mac.shellescape} netns #{q_vm}"
 
-    r "ip -n #{q_vm} tuntap add dev tap#{q_vm} mode tap user #{q_vm}"
+    r "ip netns exec #{q_vm} ip link show tap#{q_vm} > /dev/null 2>&1 || ip -n #{q_vm} tuntap add dev tap#{q_vm} mode tap user #{q_vm}"
```

Comment:
The changes here are my effort to make the function idempotent.

Reply:
Interesting. I think "r" does return a string, I wonder if you could do better. Customarily, I check error messages from `ip netns add`  (or whatever), see `rescue CommandFail` elsewhere. I think this might give you a better result than embedded shell programs, and more efficient too (fewer forks, pipes, etc)

Reply:
I changed this but didn't test it yet, I'll do so, soon.